File: compat/suites/go-sdk/internal/groups/sts.go
package groups

import (
	"context"
	"fmt"

	"github.com/Neaox/overcast-compat-go-sdk/internal/clients"
	"github.com/Neaox/overcast-compat-go-sdk/internal/harness"
	"github.com/aws/aws-sdk-go-v2/aws"
	iamsdk "github.com/aws/aws-sdk-go-v2/service/iam"
	"github.com/aws/aws-sdk-go-v2/service/sts"
)

func STS(c *clients.Clients) ServiceGroup {
	g := &stsGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"GetCallerIdentity":         g.GetCallerIdentity,
			"GetSessionToken":           g.GetSessionToken,
			"GetFederationToken":        g.GetFederationToken,
			"AssumeRole":                g.AssumeRole,
			"AssumeRoleWithWebIdentity": g.AssumeRoleWithWebIdentity,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"sts-identity": g.setupIdentity,
			"sts-assume":   g.setupAssume,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"sts-identity": g.noopTeardown,
			"sts-assume":   g.teardownAssume,
		},
	}
}

type stsGroup struct{ c *clients.Clients }

func (g *stsGroup) cl() *sts.Client { return g.c.STS() }

func (g *stsGroup) noopTeardown(_ context.Context, _ *harness.TestContext) error { return nil }

// ── sts-identity ──────────────────────────────────────────────────────────────

func (g *stsGroup) setupIdentity(_ context.Context, _ *harness.TestContext) error { return nil }

func (g *stsGroup) GetCallerIdentity(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
	if err != nil {
		return err
	}
	if aws.ToString(resp.Account) == "" {
		return fmt.Errorf("GetCallerIdentity: empty account")
	}
	return nil
}

func (g *stsGroup) GetSessionToken(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().GetSessionToken(ctx, &sts.GetSessionTokenInput{
		DurationSeconds: aws.Int32(900),
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	if resp.Credentials == nil {
		return fmt.Errorf("GetSessionToken: nil credentials")
	}
	return nil
}

func (g *stsGroup) GetFederationToken(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().GetFederationToken(ctx, &sts.GetFederationTokenInput{
		Name: aws.String("oc-fed"),
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	if resp.Credentials == nil {
		return fmt.Errorf("GetFederationToken: nil credentials")
	}
	return nil
}

// ── sts-assume ────────────────────────────────────────────────────────────────

func (g *stsGroup) setupAssume(ctx context.Context, t *harness.TestContext) error {
	// Create an IAM role to assume
	iamCl := g.c.IAM()
	roleName := fmt.Sprintf("oc-sts-%s", t.RunID)
	resp, err := iamCl.CreateRole(ctx, &iamsdk.CreateRoleInput{
		RoleName:                 aws.String(roleName),
		AssumeRolePolicyDocument: aws.String(iamAssumePolicy()),
	})
	if err != nil {
		// If IAM is not implemented, skip
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	t.Set("sts_role_name", roleName)
	t.Set("sts_role_arn", aws.ToString(resp.Role.Arn))
	return nil
}

func (g *stsGroup) teardownAssume(ctx context.Context, t *harness.TestContext) error {
	if name := t.GetString("sts_role_name"); name != "" {
		g.c.IAM().DeleteRole(ctx, &iamsdk.DeleteRoleInput{RoleName: aws.String(name)}) //nolint:errcheck
	}
	return nil
}

func (g *stsGroup) AssumeRole(ctx context.Context, t *harness.TestContext) error {
	roleArn := t.GetString("sts_role_arn")
	if roleArn == "" {
		// IAM might not be implemented; use a fake ARN and expect 501 or error
		roleArn = "arn:aws:iam::000000000000:role/oc-sts-fake"
	}
	resp, err := g.cl().AssumeRole(ctx, &sts.AssumeRoleInput{
		RoleArn:         aws.String(roleArn),
		RoleSessionName: aws.String("oc-session"),
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	if resp.Credentials == nil {
		return fmt.Errorf("AssumeRole: nil credentials")
	}
	return nil
}

func (g *stsGroup) AssumeRoleWithWebIdentity(ctx context.Context, t *harness.TestContext) error {
	roleArn := t.GetString("sts_role_arn")
	if roleArn == "" {
		roleArn = "arn:aws:iam::000000000000:role/oc-sts-fake"
	}
	resp, err := g.cl().AssumeRoleWithWebIdentity(ctx, &sts.AssumeRoleWithWebIdentityInput{
		RoleArn:          aws.String(roleArn),
		RoleSessionName:  aws.String("oc-web-session"),
		WebIdentityToken: aws.String("fake-jwt-token"),
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	if resp.Credentials == nil {
		return fmt.Errorf("AssumeRoleWithWebIdentity: nil credentials")
	}
	return nil
}

File: compat/suites/rust-sdk/src/groups/sts.rs
use std::collections::HashMap;
use std::sync::Arc;

use crate::clients::AwsClients;
use crate::groups::ServiceGroup;
use crate::harness::{TestContext, TestFn};

pub struct StsGroup {
    clients: Arc<AwsClients>,
}

impl StsGroup {
    pub fn new(clients: Arc<AwsClients>) -> Self {
        Self { clients }
    }
}

impl ServiceGroup for StsGroup {
    fn impls(&self) -> HashMap<String, TestFn> {
        let mut impls: HashMap<String, TestFn> = HashMap::new();
        let clients = self.clients.clone();
        impls.insert(
            "GetCallerIdentity".to_string(),
            Arc::new(move |_ctx: TestContext| {
                let clients = clients.clone();
                Box::pin(async move {
                    let response = clients
                        .sts()
                        .get_caller_identity()
                        .send()
                        .await
                        .map_err(|err| err.to_string())?;
                    if response.account().unwrap_or_default().is_empty() {
                        return Err("GetCallerIdentity: account missing".to_string());
                    }
                    if response.arn().unwrap_or_default().is_empty() {
                        return Err("GetCallerIdentity: arn missing".to_string());
                    }
                    if response.user_id().unwrap_or_default().is_empty() {
                        return Err("GetCallerIdentity: userId missing".to_string());
                    }
                    Ok(())
                })
            }),
        );

        let clients = self.clients.clone();
        impls.insert(
            "GetSessionToken".to_string(),
            Arc::new(move |_ctx: TestContext| {
                let clients = clients.clone();
                Box::pin(async move {
                    let response = clients
                        .sts()
                        .get_session_token()
                        .send()
                        .await
                        .map_err(|err| err.to_string())?;
                    let credentials = response
                        .credentials()
                        .ok_or_else(|| "GetSessionToken: credentials missing".to_string())?;
                    if credentials.access_key_id().is_empty() {
                        return Err("GetSessionToken: accessKeyId missing".to_string());
                    }
                    Ok(())
                })
            }),
        );

        let clients = self.clients.clone();
        impls.insert(
            "GetFederationToken".to_string(),
            Arc::new(move |_ctx: TestContext| {
                let clients = clients.clone();
                Box::pin(async move {
                    let response = clients
                        .sts()
                        .get_federation_token()
                        .name("compat-user")
                        .send()
                        .await
                        .map_err(|err| err.to_string())?;
                    let credentials = response
                        .credentials()
                        .ok_or_else(|| "GetFederationToken: credentials missing".to_string())?;
                    if credentials.access_key_id().is_empty() {
                        return Err("GetFederationToken: accessKeyId missing".to_string());
                    }
                    Ok(())
                })
            }),
        );

        let clients = self.clients.clone();
        impls.insert(
            "AssumeRole".to_string(),
            Arc::new(move |_ctx: TestContext| {
                let clients = clients.clone();
                Box::pin(async move {
                    let response = clients
                        .sts()
                        .assume_role()
                        .role_arn("arn:aws:iam::000000000000:role/compat-role")
                        .role_session_name("rust-sdk-compat")
                        .send()
                        .await
                        .map_err(|err| err.to_string())?;
                    let credentials = response
                        .credentials()
                        .ok_or_else(|| "AssumeRole: credentials missing".to_string())?;
                    if credentials.access_key_id().is_empty() {
                        return Err("AssumeRole: accessKeyId missing".to_string());
                    }
                    Ok(())
                })
            }),
        );

        let clients = self.clients.clone();
        impls.insert(
            "AssumeRoleWithWebIdentity".to_string(),
            Arc::new(move |_ctx: TestContext| {
                let clients = clients.clone();
                Box::pin(async move {
                    let response = clients
                        .sts()
                        .assume_role_with_web_identity()
                        .role_arn("arn:aws:iam::000000000000:role/compat-role")
                        .role_session_name("rust-sdk-compat")
                        .web_identity_token("fake-web-identity-token")
                        .send()
                        .await
                        .map_err(|err| err.to_string())?;
                    let credentials = response.credentials().ok_or_else(|| {
                        "AssumeRoleWithWebIdentity: credentials missing".to_string()
                    })?;
                    if credentials.access_key_id().is_empty() {
                        return Err("AssumeRoleWithWebIdentity: accessKeyId missing".to_string());
                    }
                    Ok(())
                })
            }),
        );

        impls
    }

    fn setups(&self) -> HashMap<String, TestFn> {
        HashMap::new()
    }

    fn teardowns(&self) -> HashMap<String, TestFn> {
        HashMap::new()
    }
}
